Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp

```diff
@@ -842,6 +873,91 @@ namespace Multiplayer
+            return true;
+        }
+
+        // We've received all the component data from this connection
+        // Iterate over each component and see what's been added, missing, or modified
+        for (const auto& theirComponent : theirMultiplayerComponents)
+        {
+            // Check for modified components
+            AZ::HashValue64 localComponentHash;
+            if (GetMultiplayerComponentRegistry()->FindComponentVersionHashByName(
+                    theirComponent.m_componentName, localComponentHash))
+            {
+                
+                if (theirComponent.m_componentVersionHash != localComponentHash)
+                {
+                    AZLOG_ERROR(
+                        "Multiplayer component mismatch! %s has a different version hash. Please make sure both client and server have "
+                        "matching multiplayer components.",
+                        theirComponent.m_componentName.GetCStr());
+                }
+            }
+            else
+            {
+                // Connected machine is using a multiplayer component we don't have
+                AZLOG_ERROR(
+                    "Multiplayer component mismatch! We're missing a component with version hash 0x%" PRIx64 ". "
+                    "Because we are missing this component, we don't know its name, only its hash. "
+                    "To find the missing component go to the other machine and search for 's_versionHash = AZ::HashValue64{ 0x%" PRIx64" }' "
+                    "inside the generated multiplayer auto-component build folder.",
+                    theirComponent.m_componentVersionHash,
+                    theirComponent.m_componentVersionHash);
+            }
+        }
+
+        // One last iteration over our components this time to check if we have a component the connected app is missing.
+        for (const auto& ourComponent : GetMultiplayerComponentRegistry()->GetMultiplayerComponentVersionHashes())
+        {
+            AZ::Name ourComponentName = ourComponent.first;
+
+            bool theyHaveComponent = false;
+            for (const auto& theirComponent : theirMultiplayerComponents)
+            {
+                if (ourComponentName == theirComponent.m_componentName)
+                {
+                    theyHaveComponent = true;
+                    break;
+                }
+            }
+
+            if (!theyHaveComponent)
+            {
+                AZLOG_ERROR(
+                    "Multiplayer component mismatch! We have a component named %s which the connected application is missing!",
+                    ourComponentName.GetCStr());
+            }
+        }
+
+        // Disconnect and clear out the list of components in case the user fixes up their end and tries to reconnect
+        theirMultiplayerComponents.clear();
+        connection->Disconnect(DisconnectReason::VersionMismatch, TerminationEndpoint::Local);
```

Comment:
Hmm, is there an ebus/event here that a user can sign up and detect this version mismatch? Perhaps, we should pass some extra data here as well for debug-ability?

Reply:
For example, I am curious what would be needed to expose to the Editor? Looking into the logs can often difficult.

Reply:
Yup! Events coming in the next PR
I wasn't planning anything special just for the editor, but if the mp viewport debug message system is enabled, I'll display a popup saying there's a mismatch. Also the event could be used for game-specific logic (like having the user download the latest version).